### PR TITLE
Adding Surrogate Precision@k Loss

### PR DESCRIPTION
**Pull request by AlekseySh**

### CODE REVIEW

**Comment by AlekseySh** [comment]
The loss outperformed the existing SOTA from papers with code on the SOP dataset (CMC@1 is 0.862)

File: oml/losses/surrogate_precision.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    This loss is a differentiable approximation of Precision@k metric.
+
+    The loss is described in the following paper under a bit different name:
+    `Recall@k Surrogate Loss with Large Batches and Similarity Mixup`_.
+
+    The idea is that we express the formula for Precision@k using two step functions (aka Heaviside functions).
+    Then we approximate them using two sigmoid functions with temperatures.
+    The smaller temperature the close sigmoid to the step function, but the gradients are sparser,
+    and vice versa. In the original paper `t1 = 1.0` and `t2 = 0.01` have been used.
```
Comment (DaloroAT) [inline oml/losses/surrogate_precision.py:19]: Add formula from paper?

Comment (AlekseySh) [reply oml/losses/surrogate_precision.py:19]: it's too much, link is enough

File: oml/losses/surrogate_precision.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    """
+
+    criterion_name = "surrogate_precision"
+
+    def __init__(self, k: int, temperature1: float = 1.0, temperature2: float = 0.01, reduction: str = "mean"):
+        """
+
+        Args:
+            k: Parameter
```
Comment (DaloroAT) [inline oml/losses/surrogate_precision.py:32]: some notes, that top1 always query?

**Review by DaloroAT — APPROVED** [review]

File: oml/functional/losses.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        return tensor
+
+    else:
+        ValueError(f"Unexpected type of reduction: {reduction}")
+
+
+def surrogate_precision(
+    distances: Tensor, mask_gt: Tensor, k: int, t1: float, t2: float, reduction: str = "mean"
+) -> Tensor:
+    distances_diff = distances - distances.unsqueeze(0).T
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/functional/losses.py:24]: this did not work
I went for: `distances.unsqueeze(0).permute(2, 1, 0)`

File: oml/functional/losses.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import torch
+from torch import Tensor, sigmoid
+
+
+def get_reduced(tensor: Tensor, reduction: str) -> Tensor:
+    assert reduction in ("mean", "sum", "none")
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/functional/losses.py:6]: ok

File: oml/losses/surrogate_precision.py (hunk 2 above)
Comment (AlekseySh) [reply oml/losses/surrogate_precision.py]: added

**Review by DaloroAT — APPROVED** [review]
